### DISCUSSION TITLE
How to speedup BIRT Generate Excel & PDF

### DISCUSSION THREAD

**Discussion by calvinliem** [eclipse-birt/birt#1929] [Q&A]
Hi guys, im looking for a solution. so i was comparing BIRT and JASPER with same StoreProcedure. but Jasper average generation for Excel is 500ms then BIRT is around 1.5s. Is there a tweak or config that can help BIRT faster?

Im doing compare with same Memory 2048MB between BIRT Server and JASPER Server.

**Comment by speckyspooky** [comment]
There are many questions to your compare, because you cannot compare by default with the same system settings both generator engines.
The main question would be what is the base for your testing, exactly:
- Which BIRT version was used?
- Which data would be used to generate the excel?
- Which excel type was created xls, spudsoft_xls, xlsx? 
- Can you provide the report because the report structure can speed down or speed up the genertion of excel.
- Which way was used to create the result by designer, viewer or runtime?
- Which application server and JDK was used?

There are so many dependencies.
And of course you can try to speed up the data handling with the BIRT-streaming-mode of xlsx - see: [https://github.com/eclipse-birt/birt/blob/12e0b7c733a4257e82bf974214a341ef3022efa9/engine/uk.co.spudsoft.birt.emitters.excel/src/uk/co/spudsoft/birt/emitters/excel/README.md](url)
There you will find the documentation of the user property "ExcelEmitter.StreamingXlsx" which change the data handling from full XLSX-DOM handling to streaming-output.
With the usage of the Streaming-mode not all output options will be available because these special options need the DOM-handling.

**Reply by calvinliem** [reply]
1. Which BIRT version was used? 
=> Eclipse BIRT Designer Version **4.15.0.v202403270652** & birt-runtime-**4.15.0-202403270652** running on Tomcat (dockerize)
2. Which data would be used to generate the excel?
=> PostgreSQL using JDBC Connection (postgresql-42.7.3)
3. Which excel type was created xls, spudsoft_xls, xlsx?
=> xlsx
4. Can you provide the report because the report structure can speed down or speed up the genertion of excel.
=> sure, i will drop the file below. 
5. Which way was used to create the result by designer, viewer or runtime?
=> using birt-runtime
6. Which application server and JDK was used?
=> my birtruntime running on docker with **Tomcat 9.0.91 JRE17**

[MyReport_BIRT.xlsx](https://github.com/user-attachments/files/17301420/MyReport_BIRT.xlsx)

**Comment by hvbtup** [comment]
... and do you use separate RunTask and RenderTask or do you use RunAndRenderTask?

And do you use scripting in your report?

And show us your report! 

I would expect to be JasperReports to be a little bit faster at run time, because it uses Java for logic inside the report, whereas BIRT uses Javascript.

But in my experience, most of the wallclock time when generating reports is spent in the database (or network) layer.
And I think that BIRT has advantages at development and deployment time in comparison to JasperReports (though its more than 10 years ago that I tested JasperReports).

When we switched from Oracle Reports to BIRT back then, I found that the time needed for developing reports was about the same, but runtime performance was much better (because we could use the power of Javascript), and stability was better, too.

So, while JasperReports can be faster, BIRT is usually fast enough if you develop your BIRT reports in the right way, and optimizing efforts should usually be directed at the database.

**Reply by calvinliem** [reply]
1. do you use separate RunTask and RenderTask or do you use RunAndRenderTask?
=> im not sure there's option in generate excel from birt-runtime. but im using **/run** to generate from URL
2. And do you use scripting in your report?
=> yes im using some script, in this case im using for renaming **sheetsname** cause my design is **Table inside Table then Group** (while im Grouping by country the sheetsname cannot following like if im using **Table then Group**). i will drop the structure below sir.
![image](https://github.com/user-attachments/assets/3595d5cb-71d9-4bec-a883-b057f1c5fd35)

i already upload my generated xlsx above.

yes, when im developing report design with BIRT is totally fine everything is perfect. but the birt-runtime i have tested is more slower, and i dont know what to do. im using the same StoreProcedure that i used in my JasperReport. Maybe there's birt-runtime setting to speedup generate in xlsx(?) and i've tried to generate PDF is also more terrible the gap is almost 10x times slower than jasper.

Thankyou before, im waiting for your reply sir.

**Reply by speckyspooky** [reply]
- Have you used my recommendation of the "ExcelEmitter.StreamingXlsx" optimization for xlsx?
- Structure: Why you have a table in a table (nested tables)? You should avoid this if you have a clear/simple structure it will be faster.
- Could you provide the original report as an example so we can test it?
- What is your time of PDF-generation only 10xtimes slower means Jasper 1s and BIRT 10s?

And what is your target, what makes the difference of 0.5sJasper and 1.5sBIRT
because if you have more (real) data on it or larger data rangen then you can really compare the results.
Yes it may be that the init-time can be longer on BIRT like Jasper.

Is your target to create 100.000documents per hour?

**Reply by calvinliem** [reply]
1. Have you used my recommendation of the "ExcelEmitter.StreamingXlsx" optimization for xlsx?
=> i have tried, your recommendation its get better around 300ms (rang to generate excel now in 900ms-1s). Thankyou for the recommendation.
2. Structure: Why you have a table in a table (nested tables)? You should avoid this if you have a clear/simple structure it will be faster.
=> yep, so i my first table (parent) is used for Hide or Unhide some Column as you can see i have column called "Point 1,2,3" it will show dynamically depends on the first StoreProcedure result. then i need to display the data from another StoreProcedure (different from the first one), so i need more table right? cause 1 Tabel =1 DataSet. so i wrap it using Table (Parent) inside table.
3. Could you provide the original report as an example so we can test it?
=> the .rptdesign file? maybe i need more days to send the file for you sir.
4. What is your time of PDF-generation only 10xtimes slower means Jasper 1s and BIRT 10s?
=> yes, in my case BIRT Generate PDF in 4s then jasper can do it in 400ms. (i tried to duplicate the design, cause Jasper is have Report and BIRT doesnt then i used Table inside table in this case).
5. Is your target to create 100.000documents per hour?
=> nope, i want to switch using BIRT from Jasper but the gap is too far maybe if jasper can do in 1s then birt in 1.2s it still fine. but after i tried, the time is too far. I'd love using BIRT, but in my case BIRT Runtime is much slower. so that's why im asking is there a tweak or config maybe im wrong deployed my Birt Runtime in docker or something. 

Thankyou sir, im waiting for your reply.

**Reply by hvbtup** [reply]
I'd still recommend that you find out what exactly is slow. As the very first step, you should find out if it's the DB or BIRT. 
As I said, it is the DB most of the times. And is is irrelevant if using the same DB is faster when using JasperReports, because maybe the way the SQL statements are executed differ, even if the statements might look the same in both reports.
For example, if you are using an Oracle DB, consider creating a session trace (see Oracle docs about the details how to do this) the `trcsess` and `tkprof` tools or something similar. Other RDBMS certainly provide tools to trace the session, too.
Once you can *proof* that most of the wallclock time is spent inside BIRT, go ahead and send us your *.rptdesign file (and libraries which are needed) for us to look at it.

**Reply by speckyspooky** [reply]
Have you tried for testing reason to increase the memory for BIRT (BIRT-jvm) to see may be that it could help
and here to set the -Xms / -Xmx to a higher value.

I work with many reports which includes images or I have very large data (hugh record count and column count)
and here I need more memory because otherwise I got lot of out of memory exceptions.
(This was not a problem of BIRT it was a problem of the data on my reports. More memory was helpfull.)
